primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -290,6 +292,89 @@ impl<E: Pairing> PolynomialCommitmentScheme for UnivariateKzgPCS<E> {
     }
 }
 
+impl<E, F> UnivariateKzgPCS<E>
+where
+    E: Pairing<ScalarField = F>,
+    F: FftField,
+{
+    /// Fast computation of batch opening for a single polynomial at multiple
+    /// points. Details see Sec 2.1~2.3 of [FK23](https://eprint.iacr.org/2023/033.pdf).
+    ///
+    /// Only accept `polynomial` with power-of-two degree, no constrain on the
+    /// size of `points`
+    #[allow(clippy::type_complexity)]
+    pub fn batch_open_fk23(
+        prover_param: impl Borrow<UnivariateProverParam<E>>,
+        polynomial: &<Self as PolynomialCommitmentScheme>::Polynomial,
+        points: BatchEvalPoints<E::ScalarField>,
+    ) -> Result<
+        (
+            Vec<<Self as PolynomialCommitmentScheme>::Proof>,
+            Vec<<Self as PolynomialCommitmentScheme>::Evaluation>,
+        ),
+        PCSError,
+    > {
+        let d = polynomial.degree();
+        if !d.is_power_of_two() {
+            return Err(PCSError::InvalidParameters(
+                "FK23 only supports polynomial with power-of-two degree".to_string(),
+            ));
+        }
+
+        // Step 1. compute \vec{h} using fast Toeplitz matrix multiplication
+        // 1.1 Toeplitz matrix A (named `poly_coeff_matrix` here)
+        let mut toep_col = vec![*polynomial
+            .coeffs()
+            .last()
+            .ok_or_else(|| PCSError::InvalidParameters("poly degree should >= 1".to_string()))?];
+        toep_col.resize(d, <<E as Pairing>::ScalarField as Field>::ZERO);
+        let toep_row = polynomial.coeffs().iter().skip(1).rev().cloned().collect();
+        let poly_coeff_matrix = ToeplitzMatrix::new(toep_col, toep_row)?;
+
+        // 1.2 vector s (named `srs_vec` here)
+        let srs_vec: Vec<E::G1> = prover_param
+            .borrow()
+            .powers_of_g
+            .iter()
+            .take(d)
+            .rev()
+            .cloned()
+            .map(|g| g.into_group())
+            .collect();
+
+        // 1.3 compute \vec{h}
+        let h_vec = poly_coeff_matrix.fast_vec_mul(&srs_vec)?;
+
+        // Step 2. evaluate h(X) with coeffs \vec{h} at `points` using FFT
+        let mut h_poly = GeneralDensePolynomial::from_coeff_vec(h_vec);
+        // NOTE!! this is the trickist subtle line in this function !!
+        // `polynomial` has a degree `d` which is guaranteed to be power-of-two to use
+        // Toeplitz fast mul algorithm. h(X) has a degree `d-1` which cause a plain
+        // domain instantiation on `h_poly` half the size of that for `polynomial`,
+        // which leads to different vector of roots of unity used, ultimately leading to
+        // wrong proofs. Thus, to make sure during FFT, the same vector of roots of
+        // unity are used, we need to manually resize the coefficients so that
+        // `GeneralDensePolynomial::batch_evaluate` are informed to use the domain of
+        // size (d+1).next_power_of_two().
+        h_poly.coeffs.resize(polynomial.coeffs.len(), E::G1::zero());
```

Comment:
This seems like a nasty trap for any user of `batch_evaluate`. Is there a way we can change the API to make this mistake less likely?

Reply:
Uh maybe I'm mistaken about that. Let's discuss.

Reply:
this is not something users/caller need to be careful about. This is internal subtlety for the algorithm to work.